modules/core/src/smithy4s/Blob.scala

```diff
@@ -18,101 +18,238 @@ package smithy4s
-    override def size: Int = buffer.remaining()
-    override def hashCode = buffer.hashCode()
-    override def sameBytesAs(other: Blob): Boolean = other match {
-      case otherBlob: ByteBufferBlob =>
-        buffer.compareTo(otherBlob.buffer) == 0
-      case otherBlob: ByteArrayBlob =>
-        buffer.compareTo(ByteBuffer.wrap(otherBlob.toArray)) == 0
+
+    override def asByteBufferUnsafe: ByteBuffer = buf
+
+    override def copyToBuffer(buffer: ByteBuffer, offset: Int, size: Int): Int = {
+      val toCopy = buffer.remaining.min(size)
+      buffer.put(asByteBuffer(offset, toCopy))
+      toCopy
     }
 
+    override def toString = s"ByteBufferBlob(...)"
+    override def isEmpty: Boolean = !buf.hasRemaining()
+    override def size: Int = buf.remaining()
+    override def hashCode = buf.hashCode()
+
     override def equals(other: Any): Boolean = {
       other.isInstanceOf[ByteBufferBlob] &&
-      buffer.compareTo(other.asInstanceOf[ByteBufferBlob].buffer) == 0
+      buf.compareTo(other.asInstanceOf[ByteBufferBlob].buf) == 0
     }
+  }
+
+  final class ArraySliceBlob private[smithy4s] (val arr: Array[Byte], val offset: Int, val length: Int) extends Blob {
 
-    private var arr: Array[Byte] = null
+    override def toArraySliceBlob: ArraySliceBlob = this
+
+    require(
+      offset >= 0 && offset <= arr.size && length >= 0 && length <= arr.size && offset + length <= arr.size
+    )
+    def apply(i: Int): Byte =
+      if ((offset + i) >= length) throw new IndexOutOfBoundsException()
+      else arr(offset + i)
+
+    def size: Int = length
+    def isEmpty: Boolean = (length == 0)
+
+    override def toArray: Array[Byte] = {
+      val result = Array.ofDim[Byte](length)
+      arr.copyToArray(result, offset, length)
+      result
+    }
 
-    def toArray: Array[Byte] = {
-      if (arr == null) {
-        this.synchronized {
-          if (arr == null) {
-            arr = Array.ofDim[Byte](buffer.remaining())
-            val _ = buffer.get(arr)
-          }
+    override def toArrayUnsafe: Array[Byte] = if (arr.length == length && offset == 0) arr else toArray
+
+    override def toString(): String = s"ArraySliceBlob(..., $offset, $length)"
+
+    override def hashCode(): Int = {
+      import util.hashing.MurmurHash3
+      var h = MurmurHash3.stringHash("ArraySliceBlob")
+      h = MurmurHash3.mix(h, MurmurHash3.arrayHash(arr))
+      h = MurmurHash3.mix(h, offset)
+      MurmurHash3.mixLast(h, length)
+    }
+
+    override def equals(other: Any): Boolean = {
+      other.isInstanceOf[ArraySliceBlob] && {
+        val o = other.asInstanceOf[ArraySliceBlob]
+        offset == o.offset &&
+        length == o.length &&
+        java.util.Arrays.equals(arr, o.arr)
+      }
+    }
+
+  }
+
+  final class QueueBlob private[smithy4s] (val blobs: Queue[Blob], val size: Int) extends Blob {
+    def apply(i: Int): Byte = {
+      if (i > size) throw new IndexOutOfBoundsException()
```

Comment:
```suggestion
      if (i >= size) throw new IndexOutOfBoundsException()
```

Reply:
cursor trauma in action

Reply:
Hahah I am scarred for life. This one is a much smaller issue because the `IndexOutOfBoundsException` would still get thrown, just after doing some unnecessary work.. so really a very minor optimization in this case lol

Reply:
Addressed in [aabf826](https://github.com/disneystreaming/smithy4s/pull/1211/commits/aabf8268f0b2893a451d744710eee631cd0fb68c)